src/persistent/cPersistence.c

```diff
@@ -1636,109 +1696,245 @@ simple_new(PyObject *self, PyObject *type_object)
+        return PyType_GetModuleByDef(typeobj, &CP_module_def);
+    }
 
-    if (init_strings() < 0)
+    PyErr_SetString(PyExc_TypeError, "_get_module: called w/ non-type");
+    return NULL;
+#endif
+}
+
+static cPersistenceCAPIstruct*
+_get_capi_struct(PyTypeObject *typeobj)
+{
+    PyObject* module = _get_module(typeobj);
+    if (module == NULL)
         return NULL;
 
-    module = PyModule_Create(&moduledef);
+    CP_module_state* state = PyModule_GetState(module);
+    return &(state->capi_struct);
+}
 
-    ((PyObject*)&Pertype)->ob_type = &PyType_Type;
-    Pertype.tp_new = PyType_GenericNew;
-    if (PyType_Ready(&Pertype) < 0)
+static PyObject*
+_get_copyreg_slotnames(PyTypeObject *typeobj)
+{
+    PyObject* module = _get_module(typeobj);
+    if (module == NULL)
         return NULL;
-    if (PyModule_AddObject(module, "Persistent", (PyObject *)&Pertype) < 0)
+
+    CP_module_state* state = PyModule_GetState(module);
+    return state->copyreg_slotnames;
+}
+
+static PyObject*
+_get_copyreg___newobj__(PyTypeObject *typeobj)
+{
+    PyObject* module = _get_module(typeobj);
+    if (module == NULL)
         return NULL;
 
-    cPersistenceCAPI = &truecPersistenceCAPI;
-    capi = PyCapsule_New(cPersistenceCAPI, CAPI_CAPSULE_NAME, NULL);
+    CP_module_state* state = PyModule_GetState(module);
+    return state->copyreg___newobj__;
+}
+
+static int
+CP_module_exec(PyObject* module)
+{
+    PyObject *capi;
+    PyObject *copy_reg;
+    CP_module_state* state = PyModule_GetState(module);
+    cPersistenceCAPIstruct* capi_struct = &(state->capi_struct);
+
+    /* TODO:  figure out why this can't go away, at least for heap type */
+    ((PyObject*)&Per_type_def)->ob_type = &PyType_Type;
+    Per_type_def.tp_new = PyType_GenericNew;
+
+    capi_struct->getattro = (getattrofunc)Per_getattro;
+    capi_struct->setattro = (setattrofunc)Per_setattro;
+    capi_struct->changed = changed;
+    capi_struct->accessed = accessed;
+    capi_struct->ghostify = ghostify;
+    capi_struct->setstate = (intfunctionwithpythonarg)Per_setstate,
+    /* The percachedel slot is initialized in cPickleCache.c when
+     * the module is loaded.  It uses a function in a different shared
+     * library.
+     */
+    capi_struct->percachedel = NULL,
+    capi_struct->readCurrent = readCurrent;
+
+#if USE_HEAP_ALLOCATED_TYPE
+
+    capi_struct->pertype = (PyTypeObject*)PyType_FromModuleAndSpec(
+            module, &Per_type_spec, NULL);
+
+    /* ugly hack:  'ghostify' wants to check wheter subclasses have
+     * overriden 'tp_new', and expects to be able to find the original version
+     * on .&Per_type_def'.
+     *
+     * With heap types, the version we set above gets replaced, so copy it
+     * back.
+     */
+    Per_type_def.tp_new = capi_struct->pertype->tp_new;
+
+#else
+    if (PyType_Ready(&Per_type_def) < 0)
+        return -1;
+
+    capi_struct->pertype = &Per_type_def;
+
+#endif
+
+    if (PyModule_AddObject(
```

Comment:
PyModule_AddObject steals a reference; do we need to incref for the reference in capi_struct?

Reply:
4cc597a5